### PR TITLE
jni mounting

### PR DESCRIPTION
**Pull request by tucek**
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by tucek** [comment]
Either i did it wrong, or the problem behind `kdb mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=${ELEKTRA_BUILD_DIR}/src/bindings/jna/libelektra/build/libs/libelektra-0.9.6-all.jar` resulting in 
```
construct plugin
kdb: symbol lookup error: /home/tucek/.cache/JNA/temp/jna3742494249016765927.tmp: undefined symbol: pthread_once
```
is not solved by adding `-lpthreads`

**Comment by tucek** [comment]
`kdb plugin-info -c classname=org/libelektra/plugin/Echo,classpath=/usr/local/share/java/libelektra-0.9.6-all.jar,print= jni` now results in
```
Module does not seem to be loaded.
Now in fallback code. Will directly load config from plugin.
construct plugin
open plugin
/

user:/classname
user:/classpath
user:/print
get plugin
system:/elektra/modules/jni

system:/elektra/modules/jni
system:/elektra/modules/jni/exports
system:/elektra/modules/jni/exports/close
system:/elektra/modules/jni/exports/error
system:/elektra/modules/jni/exports/get
system:/elektra/modules/jni/exports/open
system:/elektra/modules/jni/exports/set
system:/elektra/modules/jni/infos
system:/elektra/modules/jni/infos/author
system:/elektra/modules/jni/infos/description
system:/elektra/modules/jni/infos/licence
system:/elektra/modules/jni/infos/needs
system:/elektra/modules/jni/infos/placements
system:/elektra/modules/jni/infos/provides
system:/elektra/modules/jni/infos/status
system:/elektra/modules/jni/infos/version
name: system:/elektra/modules/jni/infos/provides
close plugin
/

Exported symbols: close error get open set
author: Markus Raab <elektra@libelektra.org>
description: generic Java plugin

## Introduction
[ ... README.md content ]

licence: BSD
needs:
placements: getstorage setstorage
provides: java
status: maintained unittest configurable global memleak experimental discouraged
version: 1
```
and seems to work.

`kdb mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=/usr/local/share/java/libelektra-0.9.6-all.jar` still crashes:
```
construct plugin
open plugin
/

user:/classname
user:/classpath
get plugin
system:/elektra/modules/jni

system:/elektra/modules/jni
system:/elektra/modules/jni/exports
system:/elektra/modules/jni/exports/close
system:/elektra/modules/jni/exports/error
system:/elektra/modules/jni/exports/get
system:/elektra/modules/jni/exports/open
system:/elektra/modules/jni/exports/set
system:/elektra/modules/jni/infos
system:/elektra/modules/jni/infos/author
system:/elektra/modules/jni/infos/description
system:/elektra/modules/jni/infos/licence
system:/elektra/modules/jni/infos/needs
system:/elektra/modules/jni/infos/placements
system:/elektra/modules/jni/infos/provides
system:/elektra/modules/jni/infos/status
system:/elektra/modules/jni/infos/version
name: system:/elektra/modules/jni/infos/provides
close plugin
/


Sorry, I crashed by the signal SIGSEGV
This should not have happened!

Please report the issue at https://issues.libelektra.org/
Aborted
```

but it's at least some progress

@kodebach any ideas?

File: src/tools/kdb/CMakeLists.txt [hunk 1]
```diff
@@ -21,7 +21,16 @@ if (BUILD_SHARED)
 	add_executable (kdb $<TARGET_OBJECTS:kdb-objects>)
 	add_dependencies (kdb kdb_gen_templates_generated)
 
-	target_link_libraries (kdb elektra-core elektra-kdb elektratools elektra-opts elektra-merge)
+	set (NOT_INCLUDED "")
+	plugin_check_if_included ("jni")
```
Comment (markus2330) [inline src/tools/kdb/CMakeLists.txt:26]: I don't think this is worth the hassle: there are also other plugins which might needs pthreads.

Btw. on my system, `kdb` already links with `pthread` (checked with `ldd =kdb`).

Comment (tucek) [reply src/tools/kdb/CMakeLists.txt]: for me it wasn't. updated

Comment (kodebach) [reply src/tools/kdb/CMakeLists.txt]: > Btw. on my system, `kdb` already links with `pthread` (checked with `ldd =kdb`).

Doesn't `ldd` also show libraries loaded at runtime via `dlopen`? Maybe something pulls in `pthread` dynamically, but it's not actually linked in the binary?

**Comment by kodebach** [comment]
> @kodebach any ideas?

Could you try `gdb /path/to/kdb` and then run `r mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=/usr/local/share/java/libelektra-0.9.6-all.jar`. GDB should then stop at the segfault and you can print a backtrace with `bt`. Maybe that shows where the segfault actually happens.

**Comment by tucek** [comment]
> > @kodebach any ideas?
> 
> Could you try `gdb /path/to/kdb` and then run `r mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=/usr/local/share/java/libelektra-0.9.6-all.jar`. GDB should then stop at the segfault and you can print a backtrace with `bt`. Maybe that shows where the segfault actually happens.

i fear i get the same unspecific result, you already got before:
```
tucek@Jupiter:/mnt/c/Data/libelektra-tucek/build$ gdb /usr/local/bin/kdb
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/local/bin/kdb...
(gdb) r mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=/usr/local/share/java/libelektra-0.9.6-all.jar
Starting program: /usr/local/bin/kdb mount config.jni /testsjni jni classname=org/libelektra/plugin/Echo classpath=/usr/local/share/java/libelektra-0.9.6-all.jar
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007fffdee3652d in ?? ()
(gdb) bt
#0  0x00007fffdee3652d in ?? ()
#1  0x0000000000000202 in ?? ()
#2  0x00007fffdee36624 in ?? ()
#3  0x00007ffff790b050 in ?? () from /usr/lib/jvm/java-11-openjdk-amd64/lib/server/libjvm.so
#4  0x00007fffffffc600 in ?? ()
#5  0x00007ffff750008e in ?? () from /usr/lib/jvm/java-11-openjdk-amd64/lib/server/libjvm.so
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

**Comment by tucek** [comment]
Maybe we should think about why plug-inof works, but not mount? Maybe we clean something up in the jni plugin, what should not be cleaned up?

**Comment by tucek** [comment]
or maybe @kodebach is right and we should consider the processes approach. Would this also have the benefit to allow multiple java plugins being loaded at once?

**Comment by kodebach** [comment]
> or maybe @kodebach is right and we should consider the processes approach. Would this also have the benefit to allow multiple java plugins being loaded at once?

If you use one external process per plugin I don't see why it wouldn't.

If we send the class name as well as the operation (open, get, etc.) to the Java process, you could also use a single process that dispatches to multiple plugin classes. Of course that requires a more advanced IPC method than a simple pipe, because the Java process would essentially act as a server and there would be multiple client instance on the native side.

**Comment by tucek** [comment]
working changes of this commit have been migrated to PR #3991

**Pull request closed without merging**